### PR TITLE
Add Meterpreter command compatibility requirements

### PR DESCRIPTION
**Pull request by zeroSteiner**
This is a proposed implementation to address one of the largest issues with [session compatibility validation](https://github.com/rapid7/metasploit-framework/discussions/14493) as it relates to post modules and that is dealing with the inconsistent feature set provided by the different Metepreters. This adds a `Meterpreter` key to the existing `Compat` hash which is populated with Meterpreter-compatibility information which is currently just the required commands. Using the existing module information system allows the information to be updated by mixins.

Eventually, the goal is that we would prevent post modules from running with sessions that are incompatible to pro-actively avoid issues that could confuse users. A nice bonus would be to also inform the user about why the module is incompatible and maybe even make a suggestion on how to fix it (ie migrate into a different processor architecture or spawn a different type of Meterpreter).

Compatibility checking was moved to be after the readiness check. This is necessary because the compatibility checks need to query information from Meterpreter, some of which (like the platform) isn't available until the `stdapi` extension has been loaded which is a requirement of the readiness check. At the point when compatibility is evaluated then, the `stdapi` extension is guaranteed to have been loaded. Required commands from any other extension will trigger an automatic load of the required extension in an effort to resolve the requirement. This means post modules should not have to manually load any extensions anymore as long as the compatibility requirements specify that the commands are required. Because modules shouldn't have to manually load extensions anymore, the `core_loadlib` command shouldn't be a requirement of any post modules. In the distant future, this will become important as Metasploit should support a hypothetical Meterpreter implementation that provides commands from arbitrary extensions (like `kiwi_exec_cmd` for example) without necessarily being extensible.

## Syntax
Meterpreter commands would be matched using a simple globbing pattern. This takes advantage of how commands are already named. For example, to specify that the module needs all railgun functionality, it would include `stdapi_railgun_*` as a requirement. To specify that it needs everything included in the kiwi extension, it would specify `kiwi_*` as a requirement. The compatibility checker then builds a list of all commands that match the globs and ensure that each one is provided by the session. This will need a workaround for Windows Meterpreters and their core commands per rapid7/metasploit-payloads#451.

This also leverages a new `Rex::Post::Meterpreter::CommandMapper` class.

## Limitations

- Users need to be familiar with what API calls require what Meterpreter commands.
- Modules require all functionality that a mixin requires, regardless of what methods the module uses from the mixin. See [this discussion](https://github.com/rapid7/metasploit-framework/discussions/14493#discussioncomment-637519).
- Channel support isn't covered at all. The specific channels supported by the `channel_open` command can not currently be enumerated. Moving forward, this should be supported but for now it's not a big deal since all the Meterpreters support process and file channels.

## Demonstration
For demonstration purposes I updated the `post/linux/gather/gnome_keyring_dump` module. It now declares that it's only compatible with Meterpreter sessions that offer the railgun functions, i.e. Python and not Mettle. This is in addition to the existing session type, architecture and platform checks which are confusingly left out of the `Compat` hash. The only difference before and after in the example below is the "SESSION may not be compatible with this module" message which just shows that Metasploit now knows that they're incompatible.

```
msf6 post(linux/gather/gnome_keyring_dump) > run

[!] SESSION may not be compatible with this module.
[-] Post failed: Rex::Post::Meterpreter::RequestError stdapi_railgun_api: Operation failed: The command is not supported by this Meterpreter type (x64/linux)
[-] Call stack:
[-]   /home/smcintyre/Repositories/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:213:in `send_packet_wait_response'
[-]   /home/smcintyre/Repositories/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:176:in `send_request'
[-]   /home/smcintyre/Repositories/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb:373:in `process_function_call'
[-]   /home/smcintyre/Repositories/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb:95:in `call_function'
[-]   /home/smcintyre/Repositories/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library_wrapper.rb:25:in `method_missing'
[-]   /home/smcintyre/Repositories/metasploit-framework/modules/post/linux/gather/gnome_keyring_dump.rb:196:in `run'
[*] Post module execution completed
msf6 post(linux/gather/gnome_keyring_dump) >
```

The `post/windows/gather/ntds_grabber` module was also updated. This one leverages the Powershell mixin and how the Meterpreter commands are merged in through the module info.

## Todo
The rest of the post mixins and post modules will need to be updated with their compatibility information which is why I'm opening this as a draft for discussion before I move on to that.

- [x] Handle the missing core command IDs in the Windows Meterpreter
- [x] Update the compatibility information of all the mixins
- [x] Automatically load required Meterpreter extensions
- [x] ~~Update the compatibility information of all the post modules~~
    * Going to update modules in a separate PR per an OOB suggestion

## Testing
The functionality will likely need to be spot-checked since every post-module is affected. The module requirements haven't been set yet, so modules that directly use the Meterpreter API still don't have their requirements declared (unless they overlap by chance with a required mixin).

- [x] Try running the `post/linux/gather/gnome_keyring_dump` module with a Mettle, PHP or Java Meterpreter. It should fail since only Python meets the `stdapi_railgun_*` requirement. Metasploit should warn that the sessions are incompatible (except Python of course).
- [x] Try running the `post/windows/escalate/golden_ticket` module with a fresh Windows Meterpreter. It should run after automatically loading the `kiwi` extension. Afterwards, it may fail if it's not run on a domain controller.

### CODE REVIEW

**Comment by zeroSteiner** [comment]
I didn't realize we had giant static mappings of Meterpreter command string names to command IDs in `packet.rb` when I wrote the CommandMapper class. Since it's already done though, I checked for references to these maps and didn't find any outside of `packet.rb` except for a few to `COMMAND_ID_START_CORE` which should probably be the core extension ID anyways. In commit fdf6f20 I replaced the static maps with the new dynamic CommandMapper so now we only have the commands defined in one place. The catch I guess is that the files need to be loaded.

File: lib/msf/core/post/windows/railgun.rb [hunk 1]
```diff
@@ -1,70 +0,0 @@
-# -*- coding: binary -*-
-require 'rex/post/meterpreter/extensions/stdapi/railgun/railgun'
-
-module Msf
-class Post
-module Windows
-module Railgun
```
[the remark this answers is not in the document]
Comment (space-r7) [reply lib/msf/core/post/windows/railgun.rb:7]: Only noticed that these two test modules reference this mixin:

```
test/modules/post/test/railgun.rb:  include Msf::Post::Windows::Railgun
test/modules/post/test/railgun_reverse_lookups.rb:  include Msf::Post::Windows::Railgun
```

**Review by space-r7 — COMMENTED** [review]
Looked over this, and the changes make sense to me. I just have a question for one particular mixin and a note about the removed Railgun mixin.

File: lib/msf/core/post/windows/wmic.rb [hunk 2]
```diff
@@ -11,7 +11,10 @@ module WMIC
   include Msf::Post::Windows::ExtAPI
 
   def initialize(info = {})
-    super
+    super(update_info(
+      info,
+      'Compat' => { 'Meterpreter' => { 'Commands' => %w{ stdapi_railgun_api* stdapi_sys_process_* } } }
```
[the remark this answers is not in the document]
Comment (zeroSteiner) [reply lib/msf/core/post/windows/wmic.rb:16]: Yes you're right, I missed that one. Excellent catch!

File: lib/msf/core/post/windows/railgun.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeroSteiner) [reply lib/msf/core/post/windows/railgun.rb:7]: So I tested both. `test/modules/post/test/railgun.rb` worked just fine after the missing include was removed. `test/modules/post/test/railgun_reverse_lookups.rb` needed two small functions which I added in 50c013b.

Here's the before and after with the edit in between.
```
msf6 post(test/railgun_reverse_lookups) > run

[!] SESSION may not be compatible with this module (incompatible session type: meterpreter)
[*] Running against session -1
[*] Session type is meterpreter and platform is windows
[-] FAILED: should return a constant name given a const and a filter
[-] Exception: NoMethodError : undefined method `select_const_names' for #<Msf::Modules::Post__Test__Railgun_reverse_lookups::MetasploitModule:0x0000000010cb8c38>
[-] FAILED: should return an error string given an error code
[-] Exception: NoMethodError : undefined method `lookup_error' for #<Msf::Modules::Post__Test__Railgun_reverse_lookups::MetasploitModule:0x0000000010cb8c38>
[*] Passed: 2; Failed: 0
[*] Post module execution completed
msf6 post(test/railgun_reverse_lookups) > rerun
[*] Reloading module...

[!] SESSION may not be compatible with this module (incompatible session type: meterpreter)
[*] Running against session -1
[*] Session type is meterpreter and platform is windows
[+] should return a constant name given a const and a filter
[+] should return an error string given an error code
[*] Passed: 2; Failed: 0
[*] Post module execution completed
msf6 post(test/railgun_reverse_lookups) >
```

[the remark this answers is not in the document]
Comment (space-r7) [reply lib/msf/core/post/windows/railgun.rb:7]: Worked for me:

```
msf6 exploit(multi/handler) > use post/test/railgun
msf6 post(test/railgun) > set session 1
session => 1
msf6 post(test/railgun) > run

[!] SESSION may not be compatible with this module (incompatible session type: meterpreter)
[*] Running against session 1
[*] Session type is meterpreter and platform is windows
[+] Should include error information in the results
[+] Should support functions with no parameters
[+] Should support functions with literal parameters
[+] Should support functions with in/out/inout parameter types
[+] Should support calling multiple functions at once
[+] Should support writing memory
[+] Should support reading memory
[+] Should retrieve the win32k file version
[*] Passed: 8; Failed: 0
[*] Post module execution completed
msf6 post(test/railgun) > use post/test/railgun_reverse_lookups
msf6 post(test/railgun_reverse_lookups) > set session 1
session => 1
msf6 post(test/railgun_reverse_lookups) > run

[!] SESSION may not be compatible with this module (incompatible session type: meterpreter)
[*] Running against session 1
[*] Session type is meterpreter and platform is windows
[+] should return a constant name given a const and a filter
[+] should return an error string given an error code
[*] Passed: 2; Failed: 0
[*] Post module execution completed
```

**Comment by space-r7** [comment]
Testing `post/linux/gather/gnome_keyring_dump`:

<details>
<summary>Mettle</summary>

```
msf6 exploit(multi/handler) > use post/linux/gather/gnome_keyring_dump
msf6 post(linux/gather/gnome_keyring_dump) > options

Module options (post/linux/gather/gnome_keyring_dump):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION                   yes       The session to run this module on.

msf6 post(linux/gather/gnome_keyring_dump) > set session 3
session => 3
msf6 post(linux/gather/gnome_keyring_dump) > run

[!] SESSION may not be compatible with this module (missing Meterpreter features)
[-] Post failed: Rex::Post::Meterpreter::RequestError stdapi_railgun_api: Operation failed: The command is not supported by this Meterpreter type (x64/linux)
[-] Call stack:
[-]   /Users/space/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:213:in `send_packet_wait_response'
[-]   /Users/space/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:176:in `send_request'
[-]   /Users/space/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb:373:in `process_function_call'
[-]   /Users/space/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb:95:in `call_function'
[-]   /Users/space/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library_wrapper.rb:25:in `method_missing'
[-]   /Users/space/metasploit-framework/modules/post/linux/gather/gnome_keyring_dump.rb:196:in `run'
[*] Post module execution completed
```

</details>

Getting different output for PHP:
<details>
<summary>PHP</summary>

```
msf6 exploit(multi/handler) > run

[*] Started reverse TCP handler on 192.168.37.1:4444 
[*] Sending stage (39282 bytes) to 192.168.37.131
[*] Meterpreter session 4 opened (192.168.37.1:4444 -> 192.168.37.131:53052) at 2021-05-04 14:42:36 -0500

meterpreter > sysinfo
Computer    : ubuntu
OS          : Linux ubuntu 5.8.0-50-generic #56~20.04.1-Ubuntu SMP Mon Apr 12 21:46:35 UTC 2021 x86_64
Meterpreter : php/linux
meterpreter > background
[*] Backgrounding session 4...
msf6 exploit(multi/handler) > previous
msf6 post(linux/gather/gnome_keyring_dump) > set session 4
session => 4
msf6 post(linux/gather/gnome_keyring_dump) > run

[!] SESSION may not be compatible with this module (missing Meterpreter features)
[-] Post failed: Rex::NotImplementedError The requested method is not implemented.
[-] Call stack:
[-]   /Users/space/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb:137:in `build_packet_and_layouts'
[-]   /Users/space/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb:366:in `process_function_call'
[-]   /Users/space/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb:95:in `call_function'
[-]   /Users/space/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/railgun/library_wrapper.rb:25:in `method_missing'
[-]   /Users/space/metasploit-framework/modules/post/linux/gather/gnome_keyring_dump.rb:196:in `run'
[*] Post module execution completed
```
</details>


Testing `post/windows/escalate/golden_ticket`:

<details>
<summary>Windows Meterpreter</summary>

```
msf6 exploit(multi/handler) > use post/windows/escalate/golden_ticket 
msf6 post(windows/escalate/golden_ticket) > options

Module options (post/windows/escalate/golden_ticket):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   DOMAIN                        no        Target Domain
   Domain SID                    no        Domain SID
   END_IN       87608            yes       End in ... Duration in hours, default 10 YEARS (~87608 hours)
   GROUPS                        no        ID of Groups (Comma Separated)
   ID                            no        Target User ID
   KRBTGT_HASH                   no        KRBTGT NTLM Hash
   SESSION                       yes       The session to run this module on.
   USE          false            yes       Use the ticket in the current session
   USER                          no        Target User

msf6 post(windows/escalate/golden_ticket) > set session 1
session => 1
msf6 post(windows/escalate/golden_ticket) > run

[*] Searching for the domain...
[-] Post aborted due to failure: unknown: Unable to retrieve the domain...
[*] Post module execution completed
```
</details>

**Comment by zeroSteiner** [comment]
Thanks for testing this!
> Getting different output for PHP

I'm pretty confident that this is because the Python Meterpreter [overrides](https://github.com/rapid7/metasploit-framework/blob/ee3a9e9e14b3f5047addfcdce165eea51a7670eb/lib/msf/base/sessions/meterpreter_python.rb#L109) the [`#native_arch`](https://github.com/rapid7/metasploit-framework/blob/ee3a9e9e14b3f5047addfcdce165eea51a7670eb/lib/msf/base/sessions/meterpreter.rb#L614-L621) attribute while the PHP Meterpreter does not. That means that in [`library.rb#137`](https://github.com/rapid7/metasploit-framework/blob/ee3a9e9e14b3f5047addfcdce165eea51a7670eb/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb#L137) the `arch` is "PHP". The PHP Meterpreter probably should support `#native_arch` but it hasn't been a problem in the past because it doesn't have railgun functionality so there's no need to know the real native architecture.

For the purposes of the PR here though, I think it's alright since Metasploit did correctly identify and report that the session is incompatible due to missing Meterpreter features (railgun).

**Comment by space-r7** [comment]
> I'm pretty confident that this is because the Python Meterpreter [overrides](https://github.com/rapid7/metasploit-framework/blob/ee3a9e9e14b3f5047addfcdce165eea51a7670eb/lib/msf/base/sessions/meterpreter_python.rb#L109) the [`#native_arch`](https://github.com/rapid7/metasploit-framework/blob/ee3a9e9e14b3f5047addfcdce165eea51a7670eb/lib/msf/base/sessions/meterpreter.rb#L614-L621) attribute while the PHP Meterpreter does not. That means that in [`library.rb#137`](https://github.com/rapid7/metasploit-framework/blob/ee3a9e9e14b3f5047addfcdce165eea51a7670eb/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb#L137) the `arch` is "PHP". The PHP Meterpreter probably should support `#native_arch` but it hasn't been a problem in the past because it doesn't have railgun functionality so there's no need to know the real native architecture.
> 
> For the purposes of the PR here though, I think it's alright since Metasploit did correctly identify and report that the session is incompatible due to missing Meterpreter features (railgun).

Sounds good to me! I noticed that I missed testing the `ntds_grabber` module, so here's output for that. Had to comment out a check so the railgun code would get executed, but looks like it loaded the extension and ran properly:

```
msf6 post(windows/gather/ntds_grabber) > rerun
[*] Reloading module...

[+] Running as SYSTEM
[+] PowerShell is installed.
[+] The meterpreter is the same architecture as the OS!
[*] Powershell Script executed
[*] Creating All.cab
[+] All.cab should be created in the current working directory
[*] Downloading All.cab
[+] All.cab saved in: /Users/space/.msf4/loot/20210504164308_default_192.168.37.132_CabinetFile_022469.cab
[*] Removing All.cab
[+] All.cab Removed
[*] Post module execution completed
```

**Comment by gwillcox-r7** [comment]
Leaving this to @space-r7 as she has already done most of the review on this and is in a better position to finish up the last few bits of this PR than I am.

**Review by space-r7 — COMMENTED** [review]
I was manually using the `CommandMapper` to check if the expected commands were returned for each extension, and the only extension that returned an empty hash was the `mimikatz` extension:

```
>> Rex::Post::Meterpreter::CommandMapper.get_commands('mimikatz')
=> {}
```

This is expected, though, correct? I believe Mimikatz is supported through Kiwi, but I wanted to call it out just in case. Thanks!

**Comment by zeroSteiner** [comment]
> This is expected, though, correct?
That's right, mimikatz was removed in v6, replaced by kiwi like you said. It shows up as a valid extension because it sort of is since some of the boilerplate extension code is still present however the command dispatcher was removed and that's why there are no commands.

**Comment by space-r7** [comment]
Thanks! Code looks good to me! Will go ahead and land.

**Comment by space-r7** [comment]
## Release Notes

This introduces the `meterpreter` key to the `compat` hash to better-provide Meterpreter session compatibility info to users, including printing the required Meterpreter extension(s) to the console in the event of a session incompatibility. Additionally, `post` modules will automatically load Meterpreter extensions used, provided that the module's Meterpreter compatibility requirements are annotated.

**Pull request merged**